### PR TITLE
Robust CI Restarts

### LINKED ISSUE #2060: bug/fix on false positive for link phase fail
### What is wrong?

On a user rebuild from label change in CI a corner case emerges when the link step fails and a false positive occurs because the driver scripts doesn't issue the fail to stop.

### What should have happened?

The driver script calling the build script in CI should have processed the error code correctly rather than simply exiting when a rebuild had also been issued by the user.

### What machines are impacted?

All or N/A

### Steps to reproduce

This only occurs when there is a non-zero error code returned from the link phase on a re-build request from the use from a label change.

### Additional information

Needs to be fixed so it a link failure stops the CI test so it can be addressed.

### Do you have a proposed solution?

Update the driver script to interrogate the return code to check if a link failure occurred to cover for the above corner case.

### LINKED ISSUE #2067: Add Kill Switch (Label) to CI
### What new functionality do you need?

Add a new Label MACINE-Kill that Removes PR from the CI System entirely and does not restart it.

NOTE: All the functionality for this feature is already in place with [PR 2093](https://github.com/NOAA-EMC/global-workflow/pull/2093) and should not take much programming effort.

### What are the requirements for the new functionality?

When ever a previously issued Read label PR CI test is in progress this Label update by the user will:
- Have it removed from the test suite (and data base)
- Kill the all the build scripts the CI is currently building when the kill switch is issued
- Purge the test directories from the file system
- Remove all jobs in scheduler all running from that PR

### Acceptance Criteria

Test in CI cron development environment and above capabilities can be demonstrated.

### Suggest a solution (optional)

Update the check CI bash  script to read the db state record and include a time stamp to react to this case.
This can also be used for when ever PR is merged and there are labels other than **PASSED**

### PR DESCRIPTION
**Pull request by TerrenceMcGuinness-NOAA**
# Description

Improved CI robustness for reverting back to **CI-Ready** from any given state

New Features:

- Improved `scancel` routine (refactored into bash "subroutine")
- Improved messaging (see below) when ever a user changes state
- Any and all previous build scripts and running experiments are killed as a result of reset to **Ready**

  Resolves #2060

- Bug fix (Resetting CI tests had lingering builds simultaneously running)
- New feature (Robust Resetting with messages)
- Maintenance (code refactor - use of BASH subroutines in main drivers)

Example of output when user updates state and request a restart:

```
CI Update on Hera at 11/28/23 09:33:11 PM
=================================================
PR:201 Reset to Hera-Ready by user and is now restarting CI tests
Driver PID: 302744 on hfe12 is no longer running this test
Driver PID: has restarted as 163669 on hfe12
No current experiments to cancel in PR: 201 on Hera
```

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
Reset at all steps in dev Cron CI space with appropriate kills taking place on Hera and Orion

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: ci/scripts/check_ci.sh [hunk 1]
```diff
@@ -33,6 +33,7 @@ case ${MACHINE_ID} in
 esac
 set +x
 source "${ROOT_DIR}/ush/module-setup.sh"
+source "${ROOT_DIR}/ci/scripts/utils/ci_utils.h"
```
Comment (WalterKolczynski-NOAA) [inline ci/scripts/check_ci.sh]:
```suggestion
source "${ROOT_DIR}/ci/scripts/utils/ci_utils.sh"
```

File: ci/scripts/driver.sh [hunk 2]
```diff
@@ -68,24 +69,54 @@ pr_list=$(${GH} pr list --repo "${REPO_URL}" --label "CI-${MACHINE_ID^}-Ready" -
...
+    rm -f "${output_ci_single}"
+    {
+      echo "CI Update on ${MACHINE_ID^} at $(date +'%D %r')" || true
+      echo "================================================="
+      echo "PR:${pr} Reset to ${MACHINE_ID^}-Ready by user and is now restarting CI tests" || true
+    } >> "${output_ci_single}"
+    if [[ "${driver_PID}" -ne 0 ]]; then
+      echo "Driver PID: ${driver_PID} no longer running this build having it killed"
+      if [[ "${driver_HOST}" == "${host_name}"  ]]; then
+        pstree -A -p "${driver_PID}" | grep -Eow "[0-9]+" | xargs kill || true
```
Comment (WalterKolczynski-NOAA) [inline ci/scripts/driver.sh]: It's more robust to include look-behind and look-ahead to make sure it is the right bit. Also I don't think we want to swallow any errors with `|| true`.
```suggestion
        pstree -A -p "${driver_PID}" | grep -Pow "(?<=\()[0-9]+(?=\))" | xargs kill
```

File: ci/scripts/driver.sh [hunk 3]
```diff
@@ -68,24 +69,54 @@ pr_list=$(${GH} pr list --repo "${REPO_URL}" --label "CI-${MACHINE_ID^}-Ready" -
...
+        sleep 30
```
Comment (WalterKolczynski-NOAA) [inline ci/scripts/driver.sh]: Not sure why we need to sleep.

File: ci/scripts/driver.sh [hunk 4]
```diff
@@ -68,24 +69,54 @@ pr_list=$(${GH} pr list --repo "${REPO_URL}" --label "CI-${MACHINE_ID^}-Ready" -
...
+      else
+        ssh "${driver_HOST}" 'pstree -A -p "${driver_PID}" | grep -Eow "[0-9]+" | xargs kill'
+        sleep 30
       fi
-      pslot=$(basename "${cases}")
-      sacct --format=jobid,jobname%35,WorkDir%100,stat | grep "${pslot}" | grep "PR\/${pr}\/RUNTESTS" |  awk '{print $1}' | xargs scancel || true
-    done
-    rm -Rf "${pr_dir}"
+      {
+        echo "Driver PID: ${driver_PID} on ${driver_HOST} is no longer running this test"
```
Comment (WalterKolczynski-NOAA) [inline ci/scripts/driver.sh]: Message adjustment since we can't guarantee the jobs have actually been killed (unless we check again).
```suggestion
        echo "Driver PID: Requested termination of ${driver_PID} and children on ${driver_HOST}"
```

File: ci/scripts/utils/ci_utils.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+function cancel_slurm_jobs() {
+
+  local substring=$1
+  local job_ids
+  job_ids=$(squeue -u "${USER}" -h -o "%i")
+
+  for job_id in ${job_ids}; do
+    job_name=$(sacct -j "${job_id}" --format=JobName%100 | head -3 | tail -1 | sed -r 's/\s+//g') || true
+    if [[ "${job_name}" == *"${substring}"* ]]; then
```
Comment (WalterKolczynski-NOAA) [inline ci/scripts/utils/ci_utils.sh]:
```suggestion
    if [[ "${job_name}" =~ "${substring}" ]]; then
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: ci/scripts/driver.sh [hunk 6]
```diff
@@ -68,24 +69,54 @@ pr_list=$(${GH} pr list --repo "${REPO_URL}" --label "CI-${MACHINE_ID^}-Ready" -
...
+        echo "Driver PID: has restarted as $$ on ${host_name}"
+      } >> "${output_ci_single}"
+    fi
+
+    experiments=$(find "${pr_dir}/RUNTESTS/EXPDIR" -mindepth 1 -maxdepth 1 -type d) || true
+    if [[ -z "${experiments}" ]]; then
+       echo "No current experiments to cancel in PR: ${pr} on ${MACHINE_ID^}" >> "${output_ci_single}"
+    else
+      for cases in ${experiments}; do
```
Comment (DavidHuber-NOAA) [inline ci/scripts/driver.sh]: Nitpick: should probably be `case` instead of `cases`.

File: ci/scripts/check_ci.sh (hunk 1 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/check_ci.sh]: yeah just getting to that so my jobs can now finish this morning, thanks!

File: ci/scripts/driver.sh (hunk 2 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/driver.sh]: Ok that should be fine to do a better job on the grep filter, and that darn true is for the shellnorm complaints and I need it on one line for the ssh part, I'll add an ignore exception for that then.

File: ci/scripts/driver.sh (hunk 3 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/driver.sh]: agreed, I was chasing down race conditions when I put that in there.

File: ci/scripts/utils/ci_utils.sh (hunk 5 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/utils/ci_utils.sh]: didn't know about this construct, it will still pick up substrings with this operator?

File: ci/scripts/driver.sh [hunk 7]
```diff
@@ -68,24 +69,54 @@ pr_list=$(${GH} pr list --repo "${REPO_URL}" --label "CI-${MACHINE_ID^}-Ready" -
...
+        cancel_slurm_jobs "${case_name}"
+        {
+          echo "Canceled all jobs for experiment ${case_name} in PR:${pr} on ${MACHINE_ID^}"
+        } >> "${output_ci_single}"
+      done
+    fi
+    sed -i "1 i\`\`\`" "${output_ci_single}"
+    "${GH}" pr comment "${pr}" --repo "${REPO_URL}" --body-file "${output_ci_single}"
+    db_list=$("${ROOT_DIR}/ci/scripts/pr_list_database.py" --remove_pr "${pr}" --dbfile "${pr_list_dbfile}")
+    db_list=$("${ROOT_DIR}/ci/scripts/pr_list_database.py" --add_pr "${pr}" --dbfile "${pr_list_dbfile}")
```
Comment (DavidHuber-NOAA) [inline ci/scripts/driver.sh]: I'm not sure that assigning this output to `db_list` is necessary unless some error checking needs to happen.  Should there be?

**Review by DavidHuber-NOAA — COMMENTED** [review]
Looks good, just a couple nitpicks.

File: ci/scripts/driver.sh (hunk 7 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/driver.sh]: Oh right, that was a cut and paste oversight.

File: ci/scripts/utils/ci_utils.sh (hunk 5 above)
Comment (WalterKolczynski-NOAA) [reply ci/scripts/utils/ci_utils.sh]: Yes, it is a regex matcher.

**Review by TerrenceMcGuinness-NOAA — COMMENTED** [review]
Doing extra commit to make sure reviews are in.

File: ci/scripts/driver.sh [hunk 8]
```diff
@@ -89,14 +89,14 @@ for pr in ${pr_list}; do
     if [[ "${driver_PID}" -ne 0 ]]; then
       echo "Driver PID: ${driver_PID} no longer running this build having it killed"
       if [[ "${driver_HOST}" == "${host_name}"  ]]; then
-        pstree -A -p "${driver_PID}" | grep -Eow "[0-9]+" | xargs kill || true
-        sleep 30
+        #shellcheck disable=SC2312,SC2312
```
Comment (WalterKolczynski-NOAA) [inline ci/scripts/driver.sh]:
```suggestion
        # shellcheck disable=SC2312
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: ci/scripts/driver.sh (hunk 8 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/driver.sh]: I was doubling that trying to get it to stop complaining still.

File: ci/scripts/utils/ci_utils.sh [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+  local substring=$1
+  local job_ids
+  job_ids=$(squeue -u "${USER}" -h -o "%i")
+
+  for job_id in ${job_ids}; do
+    job_name=$(sacct -j "${job_id}" --format=JobName%100 | head -3 | tail -1 | sed -r 's/\s+//g') || true
+    if [[ "${job_name}" =~ ${substring} ]]; then
+      echo "Canceling Slurm Job ${job_name} with: scancel ${job_id}"
+      scancel "${job_id}"
```
Comment (aerorahul) [inline ci/scripts/utils/ci_utils.sh:20]: You could add a `break` here since the match is made and there is nothing more to do.  Otherwise, it will just keep looping till the end of all jobs.

File: ci/scripts/driver.sh [hunk 10]
```diff
@@ -186,7 +227,7 @@ for pr in ${pr_list}; do
     done
 
     "${GH}" pr edit --repo "${REPO_URL}" "${pr}" --remove-label "CI-${MACHINE_ID^}-Building" --add-label "CI-${MACHINE_ID^}-Running"
-    "${ROOT_DIR}/ci/scripts/pr_list_database.py" --dbfile "${pr_list_dbfile}" --update_pr "${pr}" Open Running
+    "${ROOT_DIR}/ci/scripts/pr_list_database.py" --dbfile "${pr_list_dbfile}" --update_pr "${pr}" Open Running "0:0"
```
Comment (aerorahul) [inline ci/scripts/driver.sh:233]: What is "0:0"?

File: ci/scripts/driver.sh [hunk 11]
```diff
@@ -174,7 +215,7 @@ for pr in ${pr_list}; do
         } >> "${output_ci}"
       else
         {
-          echo "*** Failed *** to create experiment: ${pslot}"
+          echo "*** Failed *** to create experiment: ${pslot} on ${MACINE_ID^}"
```
Comment (aerorahul) [inline ci/scripts/driver.sh]: typo
```suggestion
          echo "*** Failed *** to create experiment: ${pslot} on ${MACHINE_ID^}"
```

File: ci/scripts/driver.sh [hunk 12]
```diff
@@ -110,34 +141,44 @@ for pr in ${pr_list}; do
...
   fi
-  "${GH}" pr edit --repo "${REPO_URL}" "${pr}" --remove-label "CI-${MACHINE_ID^}-Ready" --add-label "CI-${MACHINE_ID^}-Building"
-  "${ROOT_DIR}/ci/scripts/pr_list_database.py" --dbfile "${pr_list_dbfile}" --update_pr "${pr}" Open Building
-  echo "Processing Pull Request #${pr}"
+  id=$("${GH}" pr view "${pr}" --repo "${REPO_URL}" --json id --jq '.id')
   pr_dir="${GFS_CI_ROOT}/PR/${pr}"
+  output_ci="${pr_dir}/output_ci_${id}"
+  output_ci_single="${GFS_CI_ROOT}/PR/${pr}/output_single.log"
+  driver_build_PID=$$
+  driver_build_HOST=$(hostname -s)
```
Comment (aerorahul) [inline ci/scripts/driver.sh:152]: Isn't this already `MACHINE_ID`?

**Review by aerorahul — COMMENTED** [review]
There is a lot of dense logic being introduced here that will make it very difficult to follow (except for the developer ofcourse).
This is fine for now, but once the system is running as expected, please make a note to make it lean and clearer.

File: ci/scripts/driver.sh (hunk 12 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/driver.sh:152]: No, MACHINE_ID is say Hera, hostname is the specific head node that the PID for driver.sh is running on.

File: ci/scripts/driver.sh (hunk 10 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/driver.sh:233]: first field is PID second is `hostname`

File: ci/scripts/utils/ci_utils.sh (hunk 9 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/utils/ci_utils.sh:20]: oh yes, so is should be a `continue` so we can go on to the next one because it needs to kill all the jobs associated to that specific running Case name as it has a tag too.

File: ci/scripts/driver.sh (hunk 12 above)
Comment (aerorahul) [reply ci/scripts/driver.sh:152]: This is the exact front-end node the build is being performed.

File: ci/scripts/driver.sh (hunk 10 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/driver.sh:233]: updated documentation

File: ci/scripts/clone-build_ci.sh [hunk 13]
```diff
@@ -72,16 +72,17 @@ cd sorc || exit 1
 set +e
 ./checkout.sh -c -g -u >> log.checkout 2>&1
 checkout_status=$?
+DATE=$(date +'%D %r')
 if [[ ${checkout_status} != 0 ]]; then
   {
     echo "Checkout: *** FAILED ***"
-    echo "Checkout: Failed at $(date)" || true
+    echo "Checkout: Failed at ${DATE}" || true
```
Comment (aerorahul) [inline ci/scripts/clone-build_ci.sh]: Minor quib here: `DATE` here will be the `date` when `DATE` was set and not the actual date "now".

File: ci/scripts/driver.sh [hunk 14]
```diff
@@ -68,24 +69,57 @@ pr_list=$(${GH} pr list --repo "${REPO_URL}" --label "CI-${MACHINE_ID^}-Ready" -
...
+      echo "PR:${pr} Reset to ${MACHINE_ID^}-Ready by user and is now restarting CI tests" || true
+    } >> "${output_ci_single}"
+    if [[ "${driver_PID}" -ne 0 ]]; then
+      echo "Driver PID: ${driver_PID} no longer running this build having it killed"
+      if [[ "${driver_HOST}" == "${host_name}"  ]]; then
+        # shellcheck disable=SC2312
+        pstree -A -p "${driver_PID}" | grep -Pow "(?<=\()[0-9]+(?=\))" | xargs kill
+      else
+        # shellcheck disable=SC2312
+        ssh "${driver_HOST}" 'pstree -A -p "${driver_PID}" | grep -Eow "[0-9]+" | xargs kill'
```
Comment (aerorahul) [inline ci/scripts/driver.sh:99]: passwordless ssh in an automated script to kill jobs -- what could possibly go wrong?

**Review by TerrenceMcGuinness-NOAA — COMMENTED** [review]
updated doc and added `continue` for look for next slurm job id

File: ci/scripts/driver.sh (hunk 14 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/driver.sh:99]: I know right

Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/driver.sh:99]: My ssh-password-less was working fine on Hera and Orion is always on Login Node 1 so it will use the first clause in the IF.  When we go nems account we need to make sure ssh-passwordless-ness is working on Hera.

File: ci/scripts/clone-build_ci.sh (hunk 13 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/clone-build_ci.sh]: It's set on the line above and helps with shellnorms, I should take out the || true now.

**Review by TerrenceMcGuinness-NOAA — COMMENTED** [review]
updated all changes

File: ci/scripts/clone-build_ci.sh (hunk 13 above)
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/clone-build_ci.sh]: is in the line above and is set in to places on the condition a billionth of second later (this was to appease shell norms).

**Review by aerorahul — APPROVED** [review]
I haven't looked in detail with the changes, but as long as it works this is fine. 
We should clean this up in the future to make it clearer.

File: ci/scripts/utils/ci_utils.sh (hunk 5 above)
Comment (WalterKolczynski-NOAA) [reply ci/scripts/utils/ci_utils.sh]: Did this change get undone at some point? It is showing as `==` again.

Comment (WalterKolczynski-NOAA) [reply ci/scripts/utils/ci_utils.sh]: Updated suggestion, since shellcheck was complaining:
```suggestion
    if [[ "${job_name}" =~ ${substring} ]]; then
```

**Review by TerrenceMcGuinness-NOAA — COMMENTED** [review]
committed missed regex on BASH substring if statement.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
